tests/test_parser.py

```diff
@@ -131,3 +134,121 @@ def test_string_utf8():
     parser.consume_byte(b[:1])
     assert parser.valid_next_bytes() == set([Byte(b[1:])])
     parser.consume_byte(b[1:])
+
+
+class TestRecursiveNullableGrammars:
+    """
+    Computing parse tree of recursive nullable grammars will cause an infinite
+    loop if not handled correctly
+    """
+
+    @pytest.mark.timeout(5)
+    def test_no_infinite_loop(self):
+        """
+        A -> A
+        A ->
+
+        Loop occurs because `A -> A` is a nullable rule
```

Comment:
What happens with the JSON 'linked list' here? Doesn't that come down to something like this?

Reply:
Good question. This is pretty subtle and took me a while to understand.

The grammar
```
A -> A 'x'
A ->
```
is totally fine because the first rule isn't nullable by the definition I gave above because `'x'` isn't nullable -- only `A` is.

I.e.

`zero_or_more('x')` is fine; `zero_or_more(optional('x'))` isn't.

Linked lists look more like the former.

Reply:
I suppose the fact that the existing JSON tests passed is a good sign, but I'm still trying to work out these grammars.